src/tests/CullingTest.cpp

```diff
@@ -0,0 +1,170 @@
+    CHECK_LE(std::abs((aabbs[iBox].max() - aabbsGroundTruth[iBox].max()).dot()),
+             epsilon);
+  }
+}
+
+TEST(CullingTest, frustumCulling) {
+  // must create a GL context which will be used in the resource manager
+  esp::gfx::WindowlessContext::uptr context_ =
+      esp::gfx::WindowlessContext::create_unique(0);
+
+  // must declare these in this order due to avoid deallocation errors
+  ResourceManager resourceManager;
+  SceneManager sceneManager;
+
+  std::string sceneFile =
+      Cr::Utility::Directory::join(TEST_ASSETS, "objects/5boxes.glb");
+
+  // load the scene
+  int sceneID = sceneManager.initSceneGraph();
+  auto& sceneGraph = sceneManager.getSceneGraph(sceneID);
+  esp::scene::SceneNode& sceneRootNode = sceneGraph.getRootNode();
+  auto& drawables = sceneGraph.getDrawables();
+  const esp::assets::AssetInfo info =
+      esp::assets::AssetInfo::fromPath(sceneFile);
+  bool loadSuccess =
+      resourceManager.loadScene(info, &sceneRootNode, &drawables);
+  CHECK_EQ(loadSuccess, true);
+
+  // set the camera
+  esp::gfx::RenderCamera& renderCamera = sceneGraph.getDefaultRenderCamera();
+
+  // The camera to be set:
+  // pos: {7.3589f, -6.9258f,4.9583f}
+  // rotation: 77.4 deg, around {0.773, 0.334, 0.539}
+  // fov = 39.6 deg
+  // resolution: 800 x 600
+  // clip planes (near: 0.1m, far: 100m)
+  // with such a camera, the box 3 should be invisible, box 0, 1, 2, 4 should be
+  // visible.
+
+  // NOTE: the following test reults have been visually verified in utility
+  // viewer
+  renderCamera.setProjectionMatrix(800,     // width
+                                   600,     // height
+                                   0.01f,   // znear
+                                   100.0f,  // zfar
+                                   39.6f);  // hfov
+
+  esp::scene::SceneNode agentNode = sceneGraph.getRootNode().createChild();
+  esp::scene::SceneNode cameraNode = agentNode.createChild();
+  cameraNode.translate({7.3589f, -6.9258f, 4.9583f});
+  const Mn::Vector3 axis{0.773, 0.334, 0.539};
+  cameraNode.rotate(Mn::Math::Deg<float>(77.4f), axis.normalized());
+  renderCamera.node().setTransformation(cameraNode.absoluteTransformation());
+
+  // ground truth
+  const unsigned int visibleGroundTruth = 4;
+  const unsigned int invisibleBoxGroundTruth = 3;
+
+  // frustum culling test
+  unsigned int visible = 0;
+  Mn::Matrix4 vp =
+      renderCamera.projectionMatrix() * renderCamera.cameraMatrix();
+  const Mn::Math::Frustum<float> frustum =
+      Mn::Math::Frustum<float>::fromMatrix(vp);
+
+  for (unsigned int iDrawable = 0; iDrawable < drawables.size(); ++iDrawable) {
+    Cr::Containers::Optional<Mn::Range3D> aabb =
+        dynamic_cast<esp::scene::SceneNode&>(drawables[iDrawable].object())
+            .getAbsoluteAABB();
+    if (aabb) {
+      if (Mn::Math::Intersection::rangeFrustum(*aabb, frustum)) {
+        visible++;
+      } else {
+        CHECK_EQ(iDrawable, invisibleBoxGroundTruth);
+      }
+    }
+  }
```

Comment:
This is kinda brittle with a false sense of security because it's basically the same code as in the culling routine, but **doesn't actually call the culling routine at all**, unless I overlooked something? It's also hard to extend further.

I am not sure if testing a code by rewriting it again in a test is a good idea, because those two will inevitably get out of sync. One thing that could work is:

1. add a culling helper (`std::size_t RenderCamera::cull(std::vector<std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>, Mn::Matrix4>>&)`) that does just the `rename_if()` part, returning how many elements at front are not culled away. This one will be used from `draw()` as well as from this test.
2. get the vector of pairs from the drawable group and call `cull()` on these in the test, remember the returned value
3. set up a [SampleQuery](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1SampleQuery.html) and draw everything in the second part of the vector, expecting 0 samples to be reported by the query (to check for things that shouldn't be culled but were).
4. for each object in the first part, set up another query, clear and draw it, expect >0 samples to be drawn. Important is that each drawn object has a new query, to check for things that should be culled but weren't.
5. and as a final check, `renderCamera.draw()` should report the same value as what `cull()` returned above.

This shouldn't be too hard to set up I think, but still be pretty robust and easy to extend for new culling methods.

Reply:
OK. This is sharp. I did have a bit concern when I wrote it. Adding a new helper function seems a good idea. Will update immediately.